app/src/org/commcare/CommCareHeartbeatManager.java

```diff
@@ -0,0 +1,309 @@
+                                "Latest ccz version object in heartbeat response was not " +
+                                        "formatted properly: " + e.getMessage());
+                    }
+                }
+            });
+        } catch (SessionUnavailableException e) {
+            // Don't do anything, since we don't want to parse the response if the session service
+            // has ended.
+        }
+    }
+
+    private static void parseUpdateToPrompt(JSONObject latestVersionInfo, boolean isForApk) {
+        try {
+            if (latestVersionInfo.has("value")) {
+                String versionValue = latestVersionInfo.getString("value");
+                String forceByDate = null;
+                if (latestVersionInfo.has("force_by_date")) {
+                    forceByDate = latestVersionInfo.getString("force_by_date");
+                }
+                UpdateToPrompt updateToPrompt = new UpdateToPrompt(versionValue, forceByDate, isForApk);
+                updateToPrompt.registerWithSystem();
+            }
+        } catch (JSONException e) {
+            Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                    "Encountered malformed json while trying to parse server response into an " +
+                            "UpdateToPrompt object : " + e.getMessage());
+        }
+    }
+
+    /**
+     * @return - If the user was prompted to update
+     */
+    public static boolean promptForUpdateIfNeeded(Activity context) {
+        UpdateToPrompt cczUpdate = getCurrentUpdateToPrompt(false);
+        UpdateToPrompt apkUpdate = getCurrentUpdateToPrompt(true);
+        if (cczUpdate != null || apkUpdate != null) {
+            Intent i = new Intent(context, PromptUpdateActivity.class);
+            context.startActivity(i);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @return an UpdateToPrompt that has been stored in SharedPreferences and is still relevant
+     * (i.e. the user hasn't updated to or past this version since we stored it)
+     */
+    public static UpdateToPrompt getCurrentUpdateToPrompt(boolean forApkUpdate) {
+        CommCareApp currentApp = CommCareApplication.instance().getCurrentApp();
+        if (currentApp != null) {
+            String prefsKey = forApkUpdate ?
+                    UpdateToPrompt.KEY_APK_UPDATE_TO_PROMPT : UpdateToPrompt.KEY_CCZ_UPDATE_TO_PROMPT;
+            String serializedUpdate = currentApp.getAppPreferences().getString(prefsKey, "");
+            if (!"".equals(serializedUpdate)) {
+                try {
+                    byte[] updateBytes = Base64.decode(serializedUpdate, Base64.DEFAULT);
+                    DataInputStream stream = new DataInputStream(new ByteArrayInputStream(updateBytes));
+                    UpdateToPrompt update = (UpdateToPrompt)
+                            ExtUtil.read(stream, UpdateToPrompt.class, ExtUtil.defaultPrototypes());
```

Comment:
We have a different "master prototypes factory" set in Android that you can get from the context.